src/MainOpt.cpp

```diff
@@ -125,98 +103,50 @@ void MainOpt::parse_document(const std::string &filepath) {
-                auto const &value = broker_property.value.GetUint();
-                LOG(6, "kafka broker config {}: {}", name, value);
-                broker_opt.ConfigurationIntegers[name] = value;
-              } else {
-                LOG(3, "ERROR can not understand option: {}", name);
-              }
-            }
-          }
+  if (auto x = find<std::string>("status-uri", JSONConfiguration)) {
+    auto URIString = x.inner();
+    uri::URI URI;
+    URI.port = 9092;
+    URI.parse(URIString);
+    StatusReportURI = URI;
+  }
+}
+
+KafkaBrokerSettings
+extractKafkaBrokerSettingsFromJSON(nlohmann::json const &JSONConfiguration) {
+  KafkaBrokerSettings Settings;
+  using nlohmann::json;
+  if (auto x = find<json>("kafka", JSONConfiguration)) {
+    auto Kafka = x.inner();
+    if (auto x = find<json>("broker", Kafka)) {
+      auto Broker = x.inner();
+      for (auto Property = Broker.begin(); Property != Broker.end();
+           ++Property) {
+        auto const Key = Property.key();
+        if (Key.find("___") == 0) {
+          // Skip this property
+          continue;
+        }
+        if (Property.value().is_string()) {
+          auto Value = Property.value().get<std::string>();
+          LOG(6, "kafka broker config {}: {}", Key, Value);
+          Settings.ConfigurationStrings[Key] = Value;
+        } else if (Property.value().is_number()) {
+          auto Value = Property.value().get<int64_t>();
+          LOG(6, "kafka broker config {}: {}", Key, Value);
+          Settings.ConfigurationIntegers[Key] = Value;
+        } else {
+          LOG(3, "can not understand option: {}", Key);
         }
       }
     }
   }
-}
-
-void MainOpt::find_int(const char *key, int &property) const {
-  auto itr = json->FindMember(key);
-  if (itr != json->MemberEnd()) {
-    if (itr->value.IsInt()) {
-      property = itr->value.GetInt();
-    }
-  }
-}
-
-void MainOpt::find_main_poll_interval(int &property) {
-  find_int("main-poll-interval", property);
-}
-
-void MainOpt::find_conversion_worker_queue_size(uint32_t &property) {
-  find_uint32_t("conversion-worker-queue-size", property);
-}
-
-void MainOpt::find_uint32_t(const char *key, uint32_t &property) {
-  auto itr = json->FindMember(key);
-  if (itr != json->MemberEnd()) {
-    if (itr->value.IsInt()) {
-      property = static_cast<uint32_t>(itr->value.GetInt());
-    }
-  }
-}
-
-void MainOpt::find_conversion_threads(int &property) {
-  find_int("conversion-threads", property);
-}
-
-void MainOpt::find_broker_config(uri::URI &property) {
-  auto itr = json->FindMember("broker-config");
-  if (itr != json->MemberEnd()) {
-    if (itr->value.IsString()) {
-      property = {std::string(itr->value.GetString())};
-    }
-  }
-}
-
-void MainOpt::find_broker() {
-  auto itr = json->FindMember("broker");
-  if (itr != json->MemberEnd()) {
-    if (itr->value.IsString()) {
-      set_broker(itr->value.GetString());
-    }
-  }
+  return Settings;
```

Comment:
We can reach this point without `Settings` being initialised, would that cause problems or does it fall back on a default somewhere?

Reply:
It's not a problem. `Settings` is initialized in the first line of the function.

Reply:
It's declared there, not initialised.

Reply:
It's both declared and initialised, meaning the constructor is run and the object is in a valid state.

Reply:
Ah true, it's fine in this case, just looks wrong to me without `{}` somehow.